### PR TITLE
[CALCITE-4554] Support TIMESTAMP WITH LOCAL TIME ZONE for Snapshot and MatchRecognize

### PR DESCRIPTION
**Pull request by snuyanzin**
The PR adds support of `TIMESTAMP WITH LOCAL TIME ZONE` for Snapshot and MatchRecognize.

I added 2 `TIMESTAMP WITH LOCAL TIME ZONE` columns in `PRODUCTS_TEMPORAL` table.
It seems `MockCatalogReaderSimple#init` is pretty large since checkstyle starts complaining about method length after that. I noticed that constraint for views is the same, that's why i extracted it into a separate method and it solved issue with method length without touching checkstyle config.

### CODE REVIEW

**Comment by zoudan** [comment]
@snuyanzin we have raised a [jira](https://issues.apache.org/jira/projects/CALCITE/issues/CALCITE-5574) to split `MockCatalogReaderSimple#init` and I have created an [PR](https://github.com/apache/calcite/pull/3111) for this.

**Comment by snuyanzin** [comment]
Thanks @zoudan 
Since mentioned PR is merged https://github.com/apache/calcite/pull/3111
The current one is now rebased

**Comment by snuyanzin** [comment]
@libenchao sorry for the poke
may I ask you to have a look?
This change will allow to remove `SqlValidatorImpl` from Flink code

**Comment by libenchao** [comment]
@snuyanzin Thanks for the work, and "This change will allow to remove SqlValidatorImpl from Flink code" sounds exciting, I'll give it a review in a few days.

File: core/src/main/java/org/apache/calcite/rel/core/Snapshot.java [hunk 1]
```diff
@@ -131,7 +131,8 @@ public RexNode getPeriod() {
 
   @Override public boolean isValid(Litmus litmus, @Nullable Context context) {
     RelDataType dataType = period.getType();
-    if (dataType.getSqlTypeName() != SqlTypeName.TIMESTAMP) {
+    if (dataType.getSqlTypeName() != SqlTypeName.TIMESTAMP
```
Comment (libenchao) [inline core/src/main/java/org/apache/calcite/rel/core/Snapshot.java:134]: We can use `SqlTypeUtil#isTimestamp`?

File: core/src/test/java/org/apache/calcite/test/RelBuilderTest.java [hunk 2]
```diff
@@ -495,6 +495,30 @@ static RelBuilder createBuilder(UnaryOperator<RelBuilder.Config> transform) {
     assertThat(root, hasTree(expected));
   }
 
+  @Test void testJoinTemporalTableForTimestampWithLocalTimezone() {
+    // Equivalent SQL:
+    //   SELECT *
+    //   FROM orders
+    //   JOIN products_temporal FOR SYSTEM_TIME AS OF TIMESTAMP '2011-07-20 12:34:56'
```
Comment (libenchao) [inline core/src/test/java/org/apache/calcite/test/RelBuilderTest.java:502]:
```suggestion
    //   JOIN products_temporal FOR SYSTEM_TIME AS OF TIMESTAMP WITH LOCAL TIME ZONE '2011-07-20 12:34:56'
```

Comment (libenchao) [inline core/src/test/java/org/apache/calcite/test/RelBuilderTest.java:502]:
```suggestion
    //   JOIN products_temporal FOR SYSTEM_TIME AS OF TIMESTAMP WITH LOCAL TIME ZONE '2011-07-20 12:34:56'
```

**Review by libenchao — COMMENTED** [review]
Generally looks good, I've left some minor comments.

File: core/src/test/java/org/apache/calcite/test/RelBuilderTest.java (hunk 2 above)
Comment (snuyanzin) [reply core/src/test/java/org/apache/calcite/test/RelBuilderTest.java]: done

File: core/src/test/java/org/apache/calcite/test/SqlToRelConverterTest.java [hunk 3]
```diff
@@ -1246,6 +1246,14 @@ public static void checkActualAndReferenceFiles() {
     sql(sql).ok();
   }
 
+  @Test void testJoinTemporalTableOnSpecificTimeWithLocalTimeZone() {
```
[the remark this answers is not in the document]
Comment (snuyanzin) [reply core/src/test/java/org/apache/calcite/test/SqlToRelConverterTest.java]: done

File: core/src/main/java/org/apache/calcite/rel/core/Snapshot.java (hunk 1 above)
Comment (snuyanzin) [reply core/src/main/java/org/apache/calcite/rel/core/Snapshot.java]: yes, you are right, thanks

**Comment by snuyanzin** [comment]
Thanks for having a look

**Review by chunweilei — COMMENTED** [review]
LGTM

**Comment by snuyanzin** [comment]
@libenchao , @chunweilei , @zoudan 
thanks a lot for having a look!

**Pull request closed without merging**